### DISCUSSION TITLE
Weekly Pages

### DISCUSSION THREAD

**Discussion by stdword** [stdword/logseq13-full-house-plugin#43] [Showroom]
###### inspired by [this](https://www.youtube.com/watch?v=Q7QVVIRnKoI) Obsidian video and @inzagi1's [question](https://github.com/stdword/logseq13-full-house-plugin/discussions/42)

<img width="800px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/7e9c1c3f-14bc-4d87-8bbe-d6608ad2e45c"/>

## Steps

1) Install [`Days`](https://github.com/sethyuan/logseq-plugin-days) plugin (weekly pages available only since v0.18
   > **Important**
   > Ensure it's setting `firstWeekContainsDate` is set to 4 if your week starts with Mon, 3 if it starts with Sun and 2 if it starts with Sat.
2) Optionally install [`Another Embed`](https://github.com/sethyuan/logseq-plugin-another-embed) plugin for nice-looking embed queries.
3) Let's prepare the queries. Create anywhere in Logseq (ex: in «Queries» page) the following queries. I'll use 2 queries for \#podcast and \#diary records. But you can setup your own ones:
   ````org-mode
   reference:: #podcast
   #+BEGIN_QUERY
   …
   #+END_QUERY
   ````
   ````markdown
   reference:: #diary
   #+BEGIN_QUERY
   …
   #+END_QUERY
   ````
   Fill the body of queries with the same one code. Insert it between `#+BEGIN_QUERY` and `#+END_QUERY` with replacing the «…» (use copy button in the top right corner):
   ````clojure
   {
   :inputs [:current-page :current-block]
   :query [:find (pull ?b [*])
   :in $ ?week ?self %
   :where
   [?page :block/name ?week]
   (get-week ?page ?weekStart ?weekEnd)
   (between ?b ?weekStart ?weekEnd)

   [?self :block/page ?self-page]
   (tagged-with-property ?b ?self :reference)
   ]
   :rules [
     [(get-week ?block ?start ?end)
       [?block :block/properties ?props]
       [(get ?props :week-start) ?start]
       [(get ?props :week-end) ?end]
     ],
     [(tagged-with-property ?block ?props-block ?prop)
       [?props-block :block/properties-text-values ?props]
       [(get ?props ?prop) ?tag]
       [(subs ?tag 1) ?tag-name]

       [?block :block/refs ?ref]
       [?ref :block/name ?name]
       [(= ?name ?tag-name)]
     ]
    ]
   }
   ````
   **Restriction**: currently there is no way in Logseq to avoid copying of query body. Please vote for this feature [here](https://discuss.logseq.com/t/make-advanced-queries-reusable) and like [this](https://discuss.logseq.com/t/make-advanced-queries-reusable/15281/13?u=stdword) comment.
3) Create `🏛template` anywhere in Logseq (ex: in «Templates» page). (Use copy button in the top right corner):
   ````markdown
   - template:: weekly page
     arg-month-format:: YYYY-MM
     template-version:: 3
     plugin-version:: 4.0.0
     link:: [Weekly Pages](https://github.com/stdword/logseq13-full-house-plugin/discussions/43)
       - ```javascript
         ``{
             var page = c.page.name.match(/(\d{4}).*(?:w|W)(\d{1,2})/)
             if (!page) {
                 await logseq.UI.show_msg('This is not a weekly page')
                 return
             }
             var [_missing, year, week] = page
             
             var point = date.from(year).add(1, 'week').isoWeek(Number(week))
             var start = point.startOf('week')
             var end = point.endOf('week')
             
             var monthes = [start, end].map(d => ref(d.format(c.args['month-format']))).unique()
             outn('month:: ' + monthes.join(', '))
             outn('type:: week')
             outn('week-start:: ' + start.format('YYYYMMDD'))
             outn('week-end:: ' + end.format('YYYYMMDD'))
         }``
         ```
       - Overview
         heading:: true
           - INSERT HERE embed ref to #diary query
           - INSERT HERE embed ref to #podcast query
   ````
4) Copy reference to both of queries via right clicking on corresponding bullet and selecting «Copy block embed»:
   <img width="300px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/f43b3673-5baf-4c90-a493-e10c98309a41"/>
   …and insert them under «Overview» heading. It should look like:
   ```
       …
       - Overview
         heading:: true
           - {{embed ((666cd068-133c-4b30-9df4-1dd8a94179cb))}}
           - {{embed ((666cd1b2-f50d-4d88-859b-04f2f9e2efb2))}}
   ``` 
5) Now, let's setup `Days` plugin to automatically insert `🏛️template`. Create another template:
   ```mardown
   - template:: .weekly page inserter
     template-including-parent:: false
     - {{renderer :template, weekly page, :delay-until-rendered}}
     - 
   ```
   And add it's name to `Days` plugin's setting `weekPageTemplate`:
   <img width="489" alt="CleanShot 2024-06-17 at 10 47 06@2x" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/f0aac497-e561-463f-99a0-ce6fae3a2565">

6) Additionally you can setup [`Awesome Links`](https://github.com/yoyurec/logseq-awesome-links) plugin to have colored *references* to Weekly Pages (according to `type:: week`):
   - Goto plugin settings (`t s`) → *Awesome Links* and set the `inheritFromProp` setting to `type`:
     <img width="400" src="https://github.com/user-attachments/assets/0f854e8a-6d3b-4ad9-902b-10b37308504d">
   - Go to `[[week]]` page and add page property `color:: mediumseagreen` (choose any value you want via [this service](https://enes.in/sorted-colors)).

❤️ Enjoy the full automatic weekly page creation! 

### Why are queries embedded instead of being written directly in the template?
Because if you want to change the query, you would only need to modify it in one place - and the changes will be immediately visible on all already created weekly pages. This is very handy. But if you don't need this - you can do without _embed_ and each time "enclose" the query inside the weekly page. Then, when changing the query in the template, only new pages will see its new look.

### Why is the property `reference::` with the value `#podcast` displayed in the query block?
1. Visually, it's clear what type of data we're looking at. Although you could use the query title for this, or simply write the heading "Podcasts Overview".
2. For an easy configuration, if you suddenly want to change `#podcast` to something else (for example, to `#youtube-video`).
3. To have a block with a query in _Linked References_, on the `podcast` page. This is handy if you want to rename the `podcast` page (e.g. to `🎙️`). The block with the query will be renamed **automatically**. If it is not in _Linked References_, you will have to a) remember that the block with the query on the pages requires manual modification b) carry out this manual modification.
- If you want to hide this property, you can do this via `custom.css`:
  ```css
  .block-properties > div:has(.page-property-key[data-ref="reference"]) {
      display: none !important;
  }
  ```
  At the same time, the property continues to be visible in edit mode, which is handy if you need to change its value.

### Can I achieve the same result but without plugins?
Yes, see _Option B_ [here](https://discuss.logseq.com/t/weekly-journal-pages-auto-overview-of-the-week/27656). The disadvantage of this approach is the need to manually create links to weekly pages somewhere.
Also you can't combine Weekly Pages with [Monthly Pages](https://github.com/stdword/logseq13-full-house-plugin/discussions/44) approach.

## Additional info
- Logseq v0.10.9
- ☕️ [BuJo Theme](https://github.com/stdword/logseq-bujo-theme)
- → [discuss.logseq.com](https://discuss.logseq.com/t/weekly-journal-pages-auto-overview-of-the-week/27656)

**Comment by cmontemuino** [comment]
[Days plugin](https://github.com/sethyuan/logseq-plugin-days) is leading to `404 Page Not Found`. I guess the author has changed the project's visibility to "private"?

**Reply by stdword** [reply]
Author removes the repo. You can use [my fork](https://github.com/stdword/logseq-plugin-days).

**Reply by ard0gg** [reply]
Related: How do you get the little colored bars over the dates in the calendar?

**Reply by stdword** [reply]
> Related: How do you get the little colored bars over the dates in the calendar?

This is function of the Days plugin. Bars mark the days with non-empty journals.

**Reply by ard0gg** [reply]
Hmmm. Unfortunately that's not working out of the box on my end after building the plugin using your fork. I'll play around and see if I can figure out what's going on. I understand it's not your project and therefor not supported here.

**Reply by ard0gg** [reply]
User Error. Was using `{{renderer :days, @, all}}` instead of `{{renderer :days, *}}`.

Edit: FWIW, some themes also seem to effect this. Especially dark mode themes.

**Reply by stdword** [reply]
> some themes also seem to effect this. Especially dark mode themes.

Yes. [☕️ BuJo Theme](https://github.com/stdword/logseq-bujo-theme#readme) is totally adapted
